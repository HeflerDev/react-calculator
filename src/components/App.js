import React, { useState } from 'react';
import Display from './Display';
import ButtonPanel from './ButtonPanel';
import calculate from '../logic/calculate';

const App = () => {
  const [total, setTotal] = useState('0');
  const [next, setNext] = useState('');
  const [operation, setOperation] = useState('');
  const [reset, setReset] = useState(false);

  const setters = {
    setTotal,
    setNext,
    setOperation,
    setReset,
  };

  const getters = {
    total,
    next,
    operation,
    reset,
  };

  const handleClick = buttonName => {
    calculate(getters, setters, buttonName);
  };

  return (
    <>
      <div className="board around">
        <div className="col-12 col-m-4 col-l-7">
          <div className="content-container">
            <div className="m-auto">
              <h3>Let&apos;s make some Math!</h3>
              <p>
                The App work as a normal and functional calculator, just insert the numbers
                and calculate the results.
              </p>
            </div>
          </div>
        </div>
        <div className="col-12 col-m-8 col-l-3">
          <Display result={getters} />
          <ButtonPanel clickHandler={handleClick} />
        </div>
      </div>
    </>
  );
};

export default App;
