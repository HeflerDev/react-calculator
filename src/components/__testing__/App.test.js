import React from 'react';
import ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';
import renderer from 'react-test-renderer';
import App from '../App';

const clickOn = el => {
  act(() => {
    el.dispatchEvent(new MouseEvent('click', { bubbles: true }));
  });
};

describe('when rendering the page', () => {
  const tree = renderer
    .create(<App />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

describe('when displaying the calculator', () => {
  const btnsObj = {};
  let display;
  let board;
  let container;

  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
    act(() => {
      ReactDOM.render(<App />, container);
    });

    display = container.querySelector('.calculator-display');
    board = container.querySelector('.board');

    expect(display).toBeTruthy();
    expect(board).toBeTruthy();

    const buttons = container.querySelectorAll('.calculator-button');
    buttons.forEach(item => {
      btnsObj[item.innerHTML] = document.getElementById(`btn-${item.innerHTML}`);
    });
  });

  afterEach(() => {
    document.body.removeChild(container);
    container = null;
  });

  test('all number buttons should work', () => {
    clickOn(btnsObj['1']);
    clickOn(btnsObj['2']);
    clickOn(btnsObj['3']);
    clickOn(btnsObj['4']);
    clickOn(btnsObj['5']);
    clickOn(btnsObj['6']);
    clickOn(btnsObj['7']);
    clickOn(btnsObj['8']);
    clickOn(btnsObj['9']);
    clickOn(btnsObj['0']);
    expect(display.firstChild.innerHTML).toBe('1234567890');
  });

  test('should not stack zeroes', () => {
    clickOn(btnsObj['0']);
    clickOn(btnsObj['0']);
    clickOn(btnsObj['0']);
    expect(display.firstChild.innerHTML).toBe('0');
  });

  test('should operate plus', () => {
    clickOn(btnsObj['1']);
    clickOn(btnsObj['+']);
    clickOn(btnsObj['2']);
    clickOn(btnsObj['=']);
    expect(display.firstChild.innerHTML).toBe('3');
  });

  test('should operate minus', () => {
    clickOn(btnsObj['8']);
    clickOn(btnsObj['-']);
    clickOn(btnsObj['7']);
    clickOn(btnsObj['=']);
    expect(display.firstChild.innerHTML).toBe('1');
  });

  test('should operate times', () => {
    clickOn(btnsObj['3']);
    clickOn(btnsObj.X);
    clickOn(btnsObj['4']);
    clickOn(btnsObj['=']);
    expect(display.firstChild.innerHTML).toBe('12');
  });

  test('should operate div', () => {
    clickOn(btnsObj['1']);
    clickOn(btnsObj['0']);
    clickOn(btnsObj['/']);
    clickOn(btnsObj['5']);
    clickOn(btnsObj['=']);
    expect(display.firstChild.innerHTML).toBe('2');
  });

  test('should reset and keep working', () => {
    clickOn(btnsObj['1']);
    clickOn(btnsObj['0']);
    clickOn(btnsObj['/']);
    clickOn(btnsObj['5']);
    clickOn(btnsObj['=']);
    clickOn(btnsObj.Ac);
    expect(display.firstChild.innerHTML).toBe('0');
    clickOn(btnsObj['5']);
    expect(display.firstChild.innerHTML).toBe('5');
  });

  test('should inverse operation', () => {
    clickOn(btnsObj['1']);
    clickOn(btnsObj['2']);
    clickOn(btnsObj['3']);
    clickOn(btnsObj['+/-']);
    expect(display.firstChild.innerHTML).toBe('-123');
  });

  test('should work with decimals', () => {
    clickOn(btnsObj['.']);
    clickOn(btnsObj['3']);
    clickOn(btnsObj['3']);
    expect(display.firstChild.innerHTML).toBe('0.33');
    clickOn(btnsObj['+']);
    clickOn(btnsObj['3']);
    clickOn(btnsObj['=']);
    expect(display.firstChild.innerHTML).toBe('3.33');
  });

  test('should get %', () => {
    clickOn(btnsObj['3']);
    clickOn(btnsObj['5']);
    clickOn(btnsObj['%']);
    expect(display.firstChild.innerHTML).toBe('0.35');
  });

  test('string size should be less than 11', () => {
    for (let i = 0; i < 20; i += 1) {
      clickOn(btnsObj['5']);
    }
    expect(display.firstChild.innerHTML.length).toBeLessThanOrEqual(10);
  });

  test('should perform multiple operations', () => {
    clickOn(btnsObj['3']);
    clickOn(btnsObj['+']);
    clickOn(btnsObj['5']);
    clickOn(btnsObj['=']);
    expect(display.firstChild.innerHTML).toBe('8');
    clickOn(btnsObj['7']);
    clickOn(btnsObj.X);
    clickOn(btnsObj['7']);
    clickOn(btnsObj['=']);
    expect(display.firstChild.innerHTML).toBe('49');
    clickOn(btnsObj['3']);
    clickOn(btnsObj['0']);
    clickOn(btnsObj['5']);
    clickOn(btnsObj['0']);
    clickOn(btnsObj['-']);
    clickOn(btnsObj['1']);
    clickOn(btnsObj['5']);
    clickOn(btnsObj['0']);
    clickOn(btnsObj['0']);
    clickOn(btnsObj['=']);
    expect(display.firstChild.innerHTML).toBe('1550');
    clickOn(btnsObj['+/-']);
    expect(display.firstChild.innerHTML).toBe('-1550');
    clickOn(btnsObj['+/-']);
    expect(display.firstChild.innerHTML).toBe('1550');
    clickOn(btnsObj['%']);
    expect(display.firstChild.innerHTML).toBe('15.50');
    clickOn(btnsObj.Ac);
    clickOn(btnsObj['5']);
    clickOn(btnsObj['+']);
    clickOn(btnsObj['5']);
    clickOn(btnsObj['=']);
    expect(display.firstChild.innerHTML).toBe('10');
    clickOn(btnsObj['+']);
    clickOn(btnsObj['8']);
    clickOn(btnsObj['=']);
    expect(display.firstChild.innerHTML).toBe('18');
  });

  test('should stack operations', () => {
    clickOn(btnsObj['8']);
    clickOn(btnsObj['+']);
    clickOn(btnsObj['8']);
    clickOn(btnsObj['+']);
    expect(display.firstChild.innerHTML).toBe('16');
    clickOn(btnsObj['8']);
    clickOn(btnsObj['=']);
    expect(display.firstChild.innerHTML).toBe('24');
  });
});
