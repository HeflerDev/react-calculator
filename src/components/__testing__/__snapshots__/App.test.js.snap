// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div
  className="board around"
>
  <div
    className="col-12 col-m-4 col-l-7"
  >
    <div
      className="content-container"
    >
      <div
        className="m-auto"
      >
        <h3>
          Let's make some Math!
        </h3>
        <p>
          The App work as a normal and functional calculator, just insert the numbers and calculate the results.
        </p>
      </div>
    </div>
  </div>
  <div
    className="col-12 col-m-8 col-l-3"
  >
    <div
      className="calculator-display"
    >
      <div>
        0
      </div>
      <div>
        
      </div>
      <div>
        
      </div>
    </div>
    <div
      className="stack no-padding"
    >
      <div
        className="board"
      >
        <button
          className="calculator-button col-3"
          id="btn-Ac"
          onClick={[Function]}
          tabIndex="Ac"
          type="button"
        >
          Ac
        </button>
        <button
          className="calculator-button col-3"
          id="btn-+/-"
          onClick={[Function]}
          tabIndex="+/-"
          type="button"
        >
          +/-
        </button>
        <button
          className="calculator-button col-3"
          id="btn-%"
          onClick={[Function]}
          tabIndex="%"
          type="button"
        >
          %
        </button>
        <button
          className="calculator-button calculator-operation-button col-3"
          id="btn-/"
          onClick={[Function]}
          tabIndex="/"
          type="button"
        >
          /
        </button>
      </div>
      <div
        className="board"
      >
        <button
          className="calculator-button col-3"
          id="btn-7"
          onClick={[Function]}
          tabIndex="7"
          type="button"
        >
          7
        </button>
        <button
          className="calculator-button col-3"
          id="btn-8"
          onClick={[Function]}
          tabIndex="8"
          type="button"
        >
          8
        </button>
        <button
          className="calculator-button col-3"
          id="btn-9"
          onClick={[Function]}
          tabIndex="9"
          type="button"
        >
          9
        </button>
        <button
          className="calculator-button calculator-operation-button col-3"
          id="btn-X"
          onClick={[Function]}
          tabIndex="X"
          type="button"
        >
          X
        </button>
      </div>
      <div
        className="board"
      >
        <button
          className="calculator-button col-3"
          id="btn-4"
          onClick={[Function]}
          tabIndex="4"
          type="button"
        >
          4
        </button>
        <button
          className="calculator-button col-3"
          id="btn-5"
          onClick={[Function]}
          tabIndex="5"
          type="button"
        >
          5
        </button>
        <button
          className="calculator-button col-3"
          id="btn-6"
          onClick={[Function]}
          tabIndex="6"
          type="button"
        >
          6
        </button>
        <button
          className="calculator-button calculator-operation-button col-3"
          id="btn--"
          onClick={[Function]}
          tabIndex="-"
          type="button"
        >
          -
        </button>
      </div>
      <div
        className="board"
      >
        <button
          className="calculator-button col-3"
          id="btn-1"
          onClick={[Function]}
          tabIndex="1"
          type="button"
        >
          1
        </button>
        <button
          className="calculator-button col-3"
          id="btn-2"
          onClick={[Function]}
          tabIndex="2"
          type="button"
        >
          2
        </button>
        <button
          className="calculator-button col-3"
          id="btn-3"
          onClick={[Function]}
          tabIndex="3"
          type="button"
        >
          3
        </button>
        <button
          className="calculator-button calculator-operation-button col-3"
          id="btn-+"
          onClick={[Function]}
          tabIndex="+"
          type="button"
        >
          +
        </button>
      </div>
      <div
        className="board"
      >
        <button
          className="calculator-button col-6"
          id="btn-0"
          onClick={[Function]}
          tabIndex="0"
          type="button"
        >
          0
        </button>
        <button
          className="calculator-button col-3"
          id="btn-."
          onClick={[Function]}
          tabIndex="."
          type="button"
        >
          .
        </button>
        <button
          className="calculator-button calculator-operation-button col-3"
          id="btn-="
          onClick={[Function]}
          tabIndex="="
          type="button"
        >
          =
        </button>
      </div>
    </div>
  </div>
</div>
`;
